cilium/CFP-37315-bgp-session-with-default-gateway.md

```diff
@@ -0,0 +1,167 @@
+# CFP-37315
+
+**SIG: SIG-BGP**
+
+**Begin Design Discussion:** 2024-02-24
+
+**Cilium Release:** X.XX
+
+**Authors:** Naveen Achyuta <naveen.achyuta22@gmail.com>
+
+## Summary
+
+Cilium’s current BGP implementation requires specifying the peer IP address in the BGP cluster configuration. In large-scale environments with thousands of Kubernetes nodes, managing distinct BGP configuration files (one per node) becomes impractical.
+
+This proposal offers a simpler alternative: rely on the default gateway of ToR switches to automatically create BGP sessions when no peer IP is provided in the BGP cluster configuration.
+
+## Motivation
+
+Large networks with thousands of Kubernetes nodes cannot feasibly manage the peer IP addresses of all ToR switches within numerous BGP cluster configurations. Allowing Cilium to automatically discover and use the  default gateway as the peer simplifies BGP session creation, reducing operational overhead and configuration complexity.
+
+## Goals
+
+### Auto-Discovery of Peer IPs
+Allow Cilium to establish BGP sessions by discovering the default gateway on the node, when no peer IP is explicitly provided in the BGP cluster configuration.
+### Single Config File for All Nodes
+Enable large networks to use a generic BGP configuration file without needing to specify unique peer IP addresses on a per-node basis.
+
+## Non-Goals
+
+* _List aspects which are specifically out of context for this CFP._
+
+## Proposal
+
+### Overview
+
+At a high level, this proposal modifies how Cilium handles BGP peer IP configuration. If a user omits the peer IP for a BGP process, the Cilium agent attempts to retrieve the node’s default gateway from statedb (which stores local routing information). Cilium can then use this default gateway as the peer IP to establish the BGP session. If the default gateway is not found, the agent skips the neighbor and logs an error for it.
+
+### BGP Cluster Config
+
+If the user enables auto-discovery for a peer, cilium agent should use the discovery mode to get the peer ip. If default gateway mode is being used,
+tt should rely on statedb to get default gateway. If not found, it should skip the neighbor and log an error.
+Every peer will have to specify the address family under autoDiscovery config block so that we fetch the default gateway for that address family only.
```

Comment:
```suggestion
If the user enables auto-discovery for a peer, cilium agent should use the discovery mode to get the peer ip. If default gateway mode is being used,
it should rely on statedb to get default gateway. If not found, it should skip the neighbor and log an error.

Every peer will have to specify the address family under autoDiscovery config block so that we fetch the default gateway for that address family only.

If there are multiple default gateways, select the default gateway with the minimum metric. If it couldn't break the tie in this way, the reconciliation fails.
```